### DISCUSSION TITLE
5. ComputerVision: Facing an issue on Test_step fucntion by following your videos

### DISCUSSION THREAD

**Discussion by takvir-imran** [mrdbourke/pytorch-deep-learning#739] [General]
def test_step(model: torch.nn.Module,
    data_loader: torch.utils.data.DataLoader,
    loss_fn: torch.nn.Module,
    accuracy_fn, 
    device: torch.device = device):
  test_loss, test_acc = 0, 0
  model.eval()

  with torch.inference_mode():
    for X, y in data_loader:
      X, y = X.to(device), y.to(device)
      Xc=X.clone()
      test_pred  = model(Xc)
      test_loss += loss_fn(test_pred, y)
      test_acc += accuracy_fn(y_true=y,
                              y_pred=test_pred.argmax(dim=1))

  test_loss /= len(data_loader)
  test_acc /= len(data_loader)

  print(f"\n Test Loss: {test_loss:.5f} || Test Accuracy: {test_acc:.4f}\n")

The errror:
![Screenshot 2023-12-01 173924](https://github.com/mrdbourke/pytorch-deep-learning/assets/121662351/bb1e5cbe-18a1-42ac-95fc-1cf7fde19bbb)

@mrdbourke can you help me out. Tried different ways but facing this problem

**Comment by shisirkha** [comment]
Got ya, you need to do any modifications with the variables inside of `torch.inference_mode()` not outside of `torch.inference_mode()`
------------------------------------------------------------------------------------------------------------------------------------------------------
error:
```
#error version
for x in range(2):
   test_acc = somthing
   test_loss = something
   with torch.inference_mode():
      #testing steps here
   test_acc /= len(dataloader)
   test_loss /= len(dataloader)
```
-------------------------------------------------------------------------------------------------------------------------------------------------------
Right way:
```
for x in range(2):
   test_acc = somthing
   test_loss = something
   with torch.inference_mode():
      #testing steps here
      test_acc /= len(dataloader)
      test_loss /= len(dataloader)
```
-------------------------------------------------------------------------------------------------------------------------------------------------------
you see the difference between these examples that the right way version has an indent block(means tab space) in the test_acc and test_loss so they are inside the `torch.inference_mode():` also the error way has no indent block(means tab space) in the test_acc and test_loss variable when editing them.

so the rule is that pytorch don't allow us to edit variables outside of `torch.inference_mode():`  (That means the variables that are created in `torch.inference_mode()`)

--------------------------------------------------------------------------------------------------------------------------------------------------------
full fixed code:
```
def test_step(model: nn.Module,
              test_dataloader:  torch.utils.data.DataLoader,
              loss_fn: torch.nn.Module,
              accuracy_fn,
              device: torch.device = device):
  test_loss, test_acc = 0,0
  model = model.to(device)

  model.eval()
  with torch.inference_mode():
    for X_test, y_test in test_dataloader:
      X_test, y_test = X_test.to(device), y_test.to(device)
      test_pred = model(X_test)
      test_loss += loss_fn(test_pred,y_test)
      test_acc += accuracy_fn(y_test,test_pred.argmax(dim=1))
    #Calculate the test loss average per batch
    test_loss /= len(test_dataloader)
    #Calculate average accuracy per batch
    test_acc/=len(test_dataloader)
  #Print out what happening
  print(f"\ntest loss: {test_loss} | test acc: {test_acc}")
```
--------------------------------------------------------------------------------------------------------------------------------------------------------
Feel free to edit the code to your needs.

#LoveFromShisir
